storage/pin/pin.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pin
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"sync"
+
+	"github.com/ethersphere/swarm/api"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+)
+
+const (
+	Version            = "1.0"
+	WorkerChanSize     = 8             // Max no of goroutines when walking the file tree
+	SwarmPinHeaderName = "x-swarm-pin" // Presence of this in header indicates pinning required
+)
+
+// FileInfo is the struct that stores the information about pinned files
+// A map of this is stored in the state DB
+type FileInfo struct {
```

Comment:
FileInfo is passed as the value in the map to the state store, which checks for encoding.BinaryMarshaler interface which is not here implemented and falls back to the json.Marshal, which is, I believe is happening here. Because all methods in this file are not exported, non of this information is persisted. Do we have a test if this persistence is functional?

Reply:
Good catch. I did not update the test case for this scenario. 
Updated it now and it works.

Reply:
Cool. One other suggestion, to save a bit or two of space, to add json struct tags with a single letter value like `json:"r"`, `json:"s"`, `json:"c"`. Or if there would be a change to store every FileInfo to state store under separate keys, to add binary marshal methods. This is working just fine, I am just sharing some ideas that I got. You should decide if it is really needed.

Reply:
Well i implemented a _MarshalBinary_ and _UnmarshalBinary_ for this to be more storage efficient.